bebraven/platform
# Smarter Zoom sync logic - regenerate links when thing change

### Description
Task: https://app.asana.com/0/1201131148207877/1201271781592087

**Deploy notes**: 
* In order to support the future user stories to [delete old registrants](https://app.asana.com/0/1201131148207877/1201279111925854) (aka invalidate their link) we need to have the `registrant_id` stored on a `ZoomLinkInfo`. Also, if there aren't `ZoomLinkInfo`s populated then on the next sync the new code will re-sync (aka register) everyone. If nothing has changed, this shouldn't change the links and will result in populating the `ZoomLinkInfo`s. Since this PR doesn't invalidate the old links, the plan is just to let the sync re-sync everyone. 

* Also, deploy this first: [When there are co-LCs, Zoom prefixes should reflect both names](https://app.asana.com/0/1201131148207877/1200813128952649)

### Motivation and Context




### Test Plan
1. Make sure you have a Cohort Schedule with the "Webinar Registration X" fields configured (with different meeting IDs so you can tell if the correct link is created / updated).
    1. See the "Generate Zoom Links" tool for tips on how to create a meeting that will work with the sync.
2. Create an Enrolled Fellow and LC in the same Cohort Schedule and Cohort.
3. Create an Enrolled CP and TA.
4. Run "Sync From Salesforce" (without the Force option)
5. Make sure that the Zoom links generated on each Participant are correct and when they join, the prefixes are what we expect.
6. Update folks names and run the sync. They should be able to join using the same link and their names should be correct
    1. Note: if you do this 4 times for a single Participant, it will start to fail with a message about needing to wait a day. You can only do 3 updates per day for a given email and meeting ID pair.
7. Update folks email addresses (and their name for good measure) and run the sync. The links should change and work.
    1. Note: in this pass, the old links will also work.
    2. If you look at the registrants with the Zoom UI you'll notice that there are two for this person. One with each email. In a future pass, we'd like to delete the original one when adding a new one with a new email.
8. Update the meeting_id on the Cohort Schedule and run the sync. Folks links should change and bring them into the new meeting.
    1. Note: in this pass, the old links will still let them into the previous meeting.
9. Update the meeting password and run the sync. The links still won't work. Run it again with the "Force" option on. Now they should be updated and start working again.

### Screenshots (if appropriate):

New language in Notes a about "Force Zoom links to be updated"
![image](https://user-images.githubusercontent.com/5596986/139450102-1edbbbf6-385b-480c-bc00-c57b6b761a38.png)


### Types of changes




### Checklist:
